iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/GpuIcebergPartitioner.scala

```diff
@@ -0,0 +1,206 @@
+
+  private val partitionExprs: Seq[GpuExpression] = spec.fields().asScala.map(getPartitionExpr).toSeq
+
+  private val keyColNum: Int = spec.fields().size()
+  private val keyColIndices: Array[Int] = (0 until keyColNum).toArray
+  private val keySortOrders: Array[OrderByArg] = (0 until keyColNum)
+    .map(OrderByArg.asc(_, true))
+    .toArray
+
+  /**
+   * Partition the `input` columnar batch using iceberg's partition spec.
+   * <br/>
+   * This method takes the ownership of the input columnar batch, and it should not be used after
+   * this call.
+   */
+  def partition(input: ColumnarBatch): Seq[ColumnarBatchWithPartition] = {
+    if (input.numRows() == 0) {
+      return Seq.empty
+    }
+
+    val numRows = input.numRows()
+
+    val spillableInput = SpillableColumnarBatch(input, ACTIVE_ON_DECK_PRIORITY)
+
+    val (partitionKeys, partitions) = withRetryNoSplit(spillableInput) { scb =>
+      val parts = withResource(scb.getColumnarBatch()) { inputBatch =>
+        partitionExprs.safeMap(_.columnarEval(inputBatch))
+      }
+      val keysTable = withResource(parts) { _ =>
+        val arr = new Array[CudfColumnVector](partitionExprs.size)
+        for (i <- partitionExprs.indices) {
+          arr(i) = parts(i).getBase
+        }
+        new Table(arr:_*)
+      }
+
+      val sortedKeyTableWithRowIdx = withResource(keysTable) { _ =>
+        withResource(Scalar.fromInt(0)) { zero =>
+          withResource(CudfColumnVector.sequence(zero, numRows)) { rowIdxCol =>
+            val totalColCount = keysTable.getNumberOfColumns + 1
+            val allCols = new Array[CudfColumnVector](totalColCount)
+
+            for (i <- 0 until keysTable.getNumberOfColumns) {
+              allCols(i) = keysTable.getColumn(i)
+            }
+            allCols(keysTable.getNumberOfColumns) = rowIdxCol
+
+            withResource(new Table(allCols: _*)) { allColsTable =>
+              allColsTable.orderBy(keySortOrders: _*)
+            }
+          }
+        }
+      }
+
+      withResource(sortedKeyTableWithRowIdx) { _ =>
+        val uniqueKeysTable = sortedKeyTableWithRowIdx.groupBy(keyColIndices: _*)
+          .aggregate()
+
+        val sortedUniqueKeysTable = withResource(uniqueKeysTable) { _ =>
+          uniqueKeysTable.orderBy(keySortOrders: _*)
+        }
+
+        val (sortedPartitionKeys, splitIds) = withResource(sortedUniqueKeysTable) { _ =>
+          val partitionKeys = toPartitionKeys(spec.partitionType(),
+            partitionSparkType,
+            sortedUniqueKeysTable)
+
+          val splitIdsCv = sortedKeyTableWithRowIdx.upperBound(
+            sortedUniqueKeysTable,
+            keySortOrders: _*)
+
+          val splitIds = withResource(splitIdsCv) { _ =>
+            GpuColumnVector.toIntArray(splitIdsCv)
+          }
+
+          (partitionKeys, splitIds)
+        }
+
+        val inputTable = withResource(scb.getColumnarBatch()) { inputBatch =>
```

Comment:
Before this line, we can release useless columns in `sortedKeyTableWithRowIdx` since only the last column(row index) is needed now.